chiselFrontend/src/main/scala/chisel3/core/Module.scala

```diff
@@ -157,91 +229,160 @@ extends HasId {
+  lazy val getPorts = _firrtlPorts.get
 
-  private[chisel3] def addId(d: HasId) { _ids += d }
+  val compileOptions = moduleCompileOptions
 
-  private[core] def ports: Seq[(String,Data)] = Vector(
-    ("clock", clock), ("reset", reset), ("io", io)
-  )
+  private[core] override def generateComponent(): Component = {
+    require(!_closed, "Can't generate module more than once")
+    _closed = true
 
-  private[core] def computePorts: Seq[firrtl.Port] = {
-    // If we're auto-wrapping IO definitions, do so now.
-    if (!(compileOptions.requireIOWrap || ioDefined)) {
-      IO(io)
+    val names = nameIds(classOf[UserModule])
+
+    // Ports get first naming priority, since they are part of a Module's IO spec
+    for (port <- _ports) {
+      require(names.contains(port), s"Unable to name port $port in $this")
+      port.setRef(ModuleIO(this, _namespace.name(names(port))))
+      // Initialize output as unused
+      _commands.prepend(DefInvalid(UnlocatableSourceInfo, port.ref))
+    }
+
+    // Then everything else gets named
+    for ((node, name) <- names) {
+      node.suggestName(name)
     }
-    for ((name, port) <- ports) yield {
-      // Port definitions need to know input or output at top-level.
-      // By FIRRTL semantics, 'flipped' becomes an Input
+
+    // All suggestions are in, force names to every node.
+    for (id <- _ids) {
+      id.forceName(default="_T", _namespace)
+      id._onModuleClose
+    }
+
+    val firrtlPorts = for (port <- _ports) yield {
+      // Port definitions need to know input or output at top-level. 'flipped' means Input.
       val direction = if(Data.isFirrtlFlipped(port)) Direction.Input else Direction.Output
       firrtl.Port(port, direction)
     }
+    _firrtlPorts = Some(firrtlPorts)
+
+    val component = DefModule(this, name, firrtlPorts, _commands)
+    _component = Some(component)
+    component
   }
+}
 
-  private[core] def setupInParent(implicit sourceInfo: SourceInfo): this.type = {
-    _parent match {
-      case Some(p) => {
-        pushCommand(DefInvalid(sourceInfo, io.ref)) // init instance inputs
-        clock := override_clock.getOrElse(p.clock)
-        reset := override_reset.getOrElse(p.reset)
-        this
-      }
-      case None => this
+/** Abstract base class for Modules, which behave much like Verilog modules.
+  * These may contain both logic and state which are written in the Module
+  * body (constructor).
+  *
+  * @note Module instantiations must be wrapped in a Module() call.
+  */
+abstract class ImplicitModule()(implicit moduleCompileOptions: CompileOptions)
+    extends UserModule {
+  // Implicit clock and reset pins
+  val clock = IO(Input(Clock()))
+  val reset = IO(Input(Bool()))
+
+  // Setup ClockAndReset
+  Builder.currentClockAndReset = Some(ClockAndReset(clock, reset))
+
+  private[core] def initializeInParent() {
+    // Don't generate source info referencing parents inside a module, since this interferes with
```

Comment:
This comment doesn't make sense though. This function isn't being called IN the parent. The clock and reset connections should mean the parent has references to the child source code (which should be fine since presumably all modules are instantiating that submodule)